### PR TITLE
feat(unstable): Add Deno.Conn.ref()/unref()

### PR DESCRIPTION
**Pull request by bartlomieju**
This commit adds "Deno.Conn.ref()" and "Deno.Conn.unref()" methods.

These methods can be used to make connection block or not block the event
loop from finishing. Refing/unrefing only influences "read" operations - ie.
scheduling writes to a connection _do_ keep event loop alive.

Required for https://github.com/denoland/deno/issues/16710

### CODE REVIEW

File: ext/net/01_net.js [hunk 1]
```diff
@@ -73,6 +78,7 @@
     }
 
     read(p) {
+      // TODO(bartlomieju): handle unrefing of connection
       return read(this.rid, p);
```
Comment (bartlomieju) [inline ext/net/01_net.js:82]: Note to self: store if this conn is currently "unrefed" and if so immediately unref the promise returned from this API. Additionally all promise ids should be stored in an internal array so they all can be refed/unrefed when needed (and when the promise finishes the relevant promise id should be filtered out from that array).

**Comment by bartlomieju** [comment]
@kt3k are you able to confirm that unrefing should only influence "read" operations? This is inline with other Deno APIs, that always ensure that writes are finished and only reads can "short-circuit".

File: ext/net/01_net.js [hunk 2]
```diff
@@ -72,8 +67,28 @@
...
-    read(p) {
-      return read(this.rid, p);
+    async read(buffer) {
+      if (buffer.length === 0) {
+        return 0;
+      }
+      const promise = core.read(this.rid, buffer);
+      const promiseId = promise[promiseIdSymbol];
+      if (this.#unref) {
+        console.log("unrefed", promiseId);
```
Comment (lucacasonato) [inline ext/net/01_net.js:77]: Remove

File: ext/net/01_net.js [hunk 3]
```diff
@@ -86,7 +101,7 @@
 
     get readable() {
       if (this.#readable === undefined) {
-        this.#readable = readableStreamForRid(this.rid);
+        this.#readable = readableStreamForRidUnrefable(this.rid);
```
Comment (lucacasonato) [inline ext/net/01_net.js:104]: Right now the stream won't get unrefed if you access `.readable` after calling `.unref`. You should unref on creation based on if #unref is set.

Please also add a test for this case.

**Comment by kt3k** [comment]
@bartlomieju
> @kt3k are you able to confirm that unrefing should only influence "read" operations? This is inline with other Deno APIs, that always ensure that writes are finished and only reads can "short-circuit".

I believe that's enough for undici support.

**Review by kt3k — APPROVED** [review]
Confirmed both stream and non stream API work as intended on my machine as well.

LGTM

**Pull request merged**